### PR TITLE
cks: use HttpsURLConnection for checking api server

### PR DESCRIPTION
**Pull request by shwstppr**
### Description
Frobable fix for #4146 , #4637 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
@blueorangutan test

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/utils/KubernetesClusterUtil.java [hunk 1]
```diff
@@ -218,7 +221,10 @@ public static boolean isKubernetesClusterServerRunning(final KubernetesCluster k
         boolean k8sApiServerSetup = false;
         while (System.currentTimeMillis() < timeoutTime) {
             try {
-                String versionOutput = IOUtils.toString(new URL(String.format("https://%s:%d/version", ipAddress, port)), StringUtils.getPreferredCharset());
+                URL url = new URL(String.format("https://%s:%d/version", ipAddress, port));
+                HttpsURLConnection con = (HttpsURLConnection)url.openConnection();
```
Comment (rhtyd) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/utils/KubernetesClusterUtil.java:225]: should we configure it to ignore ssl cert validation?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by nxsbi** [comment]
> Packaging result: ✔centos7 ✖centos8 ✔debian. JID-2637

If I want to test this, how do I do so? You will have to give instructions to obtain/install the packages in my environment.

**Comment by shwstppr** [comment]
> > Packaging result: heavy_check_markcentos7 heavy_multiplication_xcentos8 heavy_check_markdebian. JID-2637
> 
> If I want to test this, how do I do so? You will have to give instructions to obtain/install the packages in my environment.

@nxsbi do you want to test the fix in 4.15 or 4.14 or it doesn't matter?

**Comment by nxsbi** [comment]
> > > Packaging result: heavy_check_markcentos7 heavy_multiplication_xcentos8 heavy_check_markdebian. JID-2637
> > 
> > 
> > If I want to test this, how do I do so? You will have to give instructions to obtain/install the packages in my environment.
> 
> @nxsbi do you want to test the fix in 4.15 or 4.14 or it doesn't matter?

@shwstppr  I want to test in 4.15.

**Comment by shwstppr** [comment]
okay @nxsbi, I'll share the 4.15 packages with the fix soon

**Comment by shwstppr** [comment]
@nxsbi please find build packages here, http://download.cloudstack.org/testing/is-4637/
They are built against current 4.15 branch.

**Comment by nxsbi** [comment]
Tried to do a quick deployment with the new packages -- 
It failed again... Below are relevant lines from Management Server log. 
Are there any additional log files I can look into or additional checks I can do? If yes, please advise.. 



`2021-02-03 17:44:42,964 INFO  [c.c.k.c.u.KubernetesClusterUtil] (API-Job-Executor-1:ctx-0078ab8c job-12195 ctx-65cc2bf7) (logid:4ac48e69) Waiting for Kubernetes cluster : k8test  master node VMs to be accessible
2021-02-03 17:44:52,965 ERROR [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-1:ctx-0078ab8c job-12195 ctx-65cc2bf7) (logid:4ac48e69) Failed to setup Kubernetes cluster : k8test3 in usable state as unable to access master node VMs of the cluster
2021-02-03 17:44:59,052 INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-1:ctx-0078ab8c job-12195 ctx-65cc2bf7) (logid:4ac48e69) Detached Kubernetes binaries from VM : k8test3-master in the Kubernetes cluster : k8test3
2021-02-03 17:45:02,995 INFO  [c.c.k.c.a.KubernetesClusterActionWorker] (API-Job-Executor-1:ctx-0078ab8c job-12195 ctx-65cc2bf7) (logid:4ac48e69) Detached Kubernetes binaries from VM : k8test3-node-1 in the Kubernetes cluster : k8test3
2021-02-03 17:45:02,998 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-1:ctx-0078ab8c job-12195) (logid:4ac48e69) Complete async job-12195, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":"530","errortext":"Failed to setup Kubernetes cluster : k8test3 in usable state as unable to access master node VMs of the cluster"}
`

**Comment by nxsbi** [comment]
@shwstppr - Tested inside Master node, with curl just to see if the certificate issue persists -- Yes it does
Note this is with the new package build you provided

image

**Comment by shwstppr** [comment]
@nxsbi can you please share output of,
`cat /var/log/cloudstack/management/management-server.log | grep -i 4ac48e69`

**Comment by nxsbi** [comment]
@shwstppr 
> cat /var/log/cloudstack/management/management-server.log | grep -i 4ac48e69

Here it is. I have cleansed the log for IP, passwords, Keys etc. so you may see XXXXX.. For IP I only changed the management LAN IPs and external IPs. You will see them as 192.168.100.XXX where XXX is the actual last bit of correct IP or  as 172.100.100.100
[mgmt-4ac48e69cleansed.log](https://github.com/apache/cloudstack/files/5924132/mgmt-4ac48e69cleansed.log)

**Comment by nxsbi** [comment]
@shwstppr 
Management server is able to curl -k the public IP of the VR. But it throws error if using just curl (without -k)

image

**Comment by shwstppr** [comment]
@nxsbi service is not being able to connect to SSH with VR public IP and forwarded port (must be 2222), failing here
https://github.com/apache/cloudstack/blob/280c13a4bb103dd748ec304bfe0714a148c24602/plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/utils/KubernetesClusterUtil.java#L247-L253

Are you able to manually do SSH from the management server?
`ssh -i /root/.ssh/id_rsa.cloud -p 2222 core@public_ip_of_network`

Firewall and port forwarding rules must be automatically provisioned by the service in the cluster's network.
Service uses SSH over 2222 to 2222+n ports. SSH on worker nodes are done only during k8s version upgrade

**Comment by nxsbi** [comment]
@shwstppr That was it!! The Management Server was not able to reach the Virtual router on port 2222 via ssh. 
Turned out, the main Hardware firewall (outside of Cloudstack) does not allow inter VLAN communication as a default. That is what was blocking it. I added rule to allow the communication, and was able to create a cluster just now! 

Thanks so much!

**Comment by nxsbi** [comment]
Is there any option for Verbose logging which would have logged the ipAddress and Port that were used in the? That would have made it easier to determine the problem of not being able to connect to VR IP on port 2222... Just a thought. 

`
socket.connect(new InetSocketAddress(ipAddress, port), 10000)`

**Comment by shwstppr** [comment]
@nxsbi there is no way to make logging verbose for that section of code. Maybe something to improve in next release

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
LGTM

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/utils/KubernetesClusterUtil.java (hunk 1 above)
Comment (DaanHoogland) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/utils/KubernetesClusterUtil.java:232]: and get more static analysis reports as issues?

**Comment by rhtyd** [comment]
@shwstppr is this good to go?

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by shwstppr** [comment]
@

**Comment by shwstppr** [comment]
@weizhouapache possible for you to test this and see if this fixes curl -L issue you highlighted in #4146

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Review by weizhouapache — APPROVED** [review]

**Comment by rhtyd** [comment]
@blueorangutan test

**Pull request merged**